tests/client.rs

```diff
@@ -0,0 +1,146 @@
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+
+    fn get_addr(node_id: u64, cluster_id: u64) -> String {
+        format!("127.0.0.1:{}", 31000 + cluster_id * 100 + node_id)
+    }
+    
+    fn get_rpc_addr(node_id: u64, cluster_id: u64) -> String {
+        format!("127.0.0.1:{}", 32000 + cluster_id * 100 + node_id)
+    }
+
+    // We can define these later in a Config.toml
+    let num_clusters = 3;
+    let nodes_per_cluster = 3;
+
+    let mut all_nodes = Vec::new();
+    let mut node_map = HashMap::new(); // Map to store address to RaftNode mapping
+
+    // Start clusters
+    for cluster_id in 1..=num_clusters {
+        let mut cluster_nodes = Vec::new();
+        let temp_dirs: Vec<_> = (1..=nodes_per_cluster)
+            .map(|_| tempfile::TempDir::new().unwrap())
+            .collect();
+
+        let handle = Handle::current();
+        for node_id in 1..=nodes_per_cluster {
+            let handle_clone = handle.clone();
+            let temp_dir = temp_dirs[node_id as usize - 1].path().to_path_buf();
+            let addr = get_addr(node_id, cluster_id);
+            let rpc_addr = get_rpc_addr(node_id, cluster_id);
+            let addr_clone = addr.clone();
+            thread::spawn(move || {
+                let x = handle_clone.block_on(start_example_raft_node(
+                    node_id,
+                    &temp_dir,
+                    addr_clone,
+                    rpc_addr,
+                ));
+                println!("x: {:?}", x);
+            });
+            cluster_nodes.push((addr, 1.0));
+        }
+        all_nodes.push(cluster_nodes);
+    }
+
+    // Wait for servers to start up.
+    tokio::time::sleep(Duration::from_millis(1_000)).await;
+
+    // Initialize each cluster
+    for (cluster_id, nodes) in all_nodes.iter().enumerate() {
+        let leader = RaftNode::new(1, nodes[0].0.clone());
+        println!("=== init cluster {} with leader at {}", cluster_id + 1, nodes[0].0);
+        leader.init().await?;
+        for (node_id, node) in nodes.iter().enumerate().skip(1) {
+            println!("=== add node {} to cluster {}", node_id + 1, cluster_id + 1);
+            leader.add_learner((node_id as u64 + 1, node.0.clone(), get_rpc_addr(node_id as u64 + 1, cluster_id as u64 + 1))).await?;
+        }
+        println!("=== change-membership for cluster {}", cluster_id + 1);
+        leader.change_membership(&nodes.iter().enumerate().map(|(id, _)| id as u64 + 1).collect()).await?;
+    
+        node_map.insert(nodes[0].0.clone(), leader);
+    }
+
+    // Create a CARP ring with the initial leaders
+    // TODO: this needs to be changed bc leaders change
+    let initial_load = 1.0 / num_clusters as f32;
+    let mut carp_ring = Carp::new(
+        all_nodes.iter().map(|cluster| {
+            let (leader_addr, _) = &cluster[0];
+            (leader_addr.clone(), initial_load)
+        }).collect(),
+        0,
+    );
+
+    // Display the CARP ring nodes
```

Comment:
We could implement the Display trait for Carp in `carp.rs` if you want to print it out

Reply:
will do!